File: backend/src/db/models/Problem.ts
import { Schema, model } from 'mongoose';

export interface ITestCase {
  id: string;
  input: string;
  expectedOutput: string;
  isSample: boolean;
  sortOrder: number;
}

export interface IProblem {
  _id: string;
  title: string;
  slug: string;
  difficulty: 'easy' | 'medium' | 'hard';
  points: number;
  timeLimitMs: number;
  memoryLimitMb: number;
  acceptanceRate: number;
  totalSubmissions: number;
  description: string;
  inputFormat: string;
  outputFormat: string;
  tags: string[];
  isActive: boolean;
  createdBy?: string;
  testCases: ITestCase[];
  createdAt: Date;
}

const testCaseSchema = new Schema<ITestCase>(
  {
    id: { type: String, required: true },
    input: { type: String, default: '' },
    expectedOutput: { type: String, default: '' },
    isSample: { type: Boolean, default: false },
    sortOrder: { type: Number, default: 0 },
  },
  { _id: false },
);

const problemSchema = new Schema<IProblem>(
  {
    _id: { type: String, required: true },
    title: { type: String, required: true },
    slug: { type: String, required: true, unique: true, index: true },
    difficulty: {
      type: String,
      enum: ['easy', 'medium', 'hard'],
      required: true,
    },
    points: { type: Number, default: 100 },
    timeLimitMs: { type: Number, default: 1000 },
    memoryLimitMb: { type: Number, default: 256 },
    acceptanceRate: { type: Number, default: 0 },
    totalSubmissions: { type: Number, default: 0 },
    description: { type: String, default: '' },
    inputFormat: { type: String, default: '' },
    outputFormat: { type: String, default: '' },
    tags: [{ type: String }],
    isActive: { type: Boolean, default: true },
    createdBy: { type: String, ref: 'User' },
    testCases: [testCaseSchema],
    createdAt: { type: Date, default: Date.now },
  },
  {
    timestamps: false,
    versionKey: false,
  },
);

export const Problem = model<IProblem>('Problem', problemSchema);

File: backend/app/schemas/problem.py
from pydantic import BaseModel, Field
from typing import List, Literal, Optional

class TestCaseSchema(BaseModel):
    id: Optional[str] = None
    input: str = ""
    expectedOutput: str = Field(..., alias="expected_output")
    isSample: bool = Field(default=False, alias="is_sample")
    sortOrder: int = Field(default=0, alias="sort_order")

    class Config:
        populate_by_name = True

class ProblemCreate(BaseModel):
    title: str = Field(..., min_length=2)
    difficulty: Literal["easy", "medium", "hard"]
    points: int = 100
    timeLimitMs: int = Field(default=1000, alias="time_limit_ms")
    memoryLimitMb: int = Field(default=256, alias="memory_limit_mb")
    description: str = ""
    inputFormat: str = Field(default="", alias="input_format")
    outputFormat: str = Field(default="", alias="output_format")
    tags: List[str] = Field(default_factory=list)
    testCases: List[TestCaseSchema] = Field(default_factory=list, alias="test_cases")

    class Config:
        populate_by_name = True

class ProblemResponse(BaseModel):
    id: str = Field(..., alias="id")
    title: str
    slug: str
    difficulty: str
    points: int
    timeLimitMs: int = Field(..., alias="time_limit_ms")
    memoryLimitMb: int = Field(..., alias="memory_limit_mb")
    acceptanceRate: float = Field(..., alias="acceptance_rate")
    totalSubmissions: int = Field(..., alias="total_submissions")
    description: str
    inputFormat: str = Field(..., alias="input_format")
    outputFormat: str = Field(..., alias="output_format")
    tags: List[str]
    isActive: bool = Field(..., alias="is_active")
    testCases: List[TestCaseSchema] = Field(..., alias="test_cases")

    class Config:
        populate_by_name = True
